src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -4768,6 +4811,62 @@ internal static Collection<string> GetResolvedPathCollection(string filePath, Ex
             return filePaths;
         }
 
+        internal PSSession GetWindowsPowerShellCompatRemotingSession()
+        {
+            PSSession result = null;
+            var commandInfo = new CmdletInfo("Get-PSSession", typeof(GetPSSessionCommand));
+            using var ps = System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace);
+            ps.AddCommand(commandInfo);
+            ps.AddParameter("Name", WindowsPowerShellCompatRemotingSessionName);
+            var results = ps.Invoke<PSSession>();
+            if (results.Count > 0)
+            {
+                result = results[0];
+            }
+            return result;
+        }
+
+        internal PSSession CreateWindowsPowerShellCompatResources()
+        {
+            PSSession compatSession = null;
+            lock(WindowsPowerShellCompatSyncObject)
```

Comment:
* "How many times will one process need to create a new compat session?" - may be multiple times. Compat session is removed in code if the last module that is using it is removed; it also can be removed externally (by user); in either case we need to recreate the compat session next time when compat-module is loaded.
* "What's the conceptual model around this lock?". Basically by design a compat session is a singleton-type of resource shared across the entire process. Creation and deletion of such shared resource should be synchronized. Same idea about `s_WindowsPowerShellCompatUsageCounter` counter - it may be accessed and modified from multiple threads, so access to it is synchronized with `Interlocked.Exchange`.
* "Can't we just keep the value cached as a static?" That is how it was done originally :) However testing showed that some scenarios break in this case (especially when session is removed externally).

Reply:
> Basically by design a compat session is a singleton-type of resource shared across the entire process. Creation and deletion of such shared resource should be synchronized. Same idea about s_WindowsPowerShellCompatUsageCounter counter - it may be accessed and modified from multiple threads, so access to it is synchronized with Interlocked.Exchange.

That part I understand, but that's the conceptual model around the shared session, rather than the lock -- the lock isn't used everywhere we interact with the compat session, so its purpose is more nuanced than that. Anyway, I suppose for now the lock at least guarantees safety, even if it's probably too broad.